ledger/src/bank_forks.rs

```diff
@@ -177,34 +177,54 @@ impl BankForks {
             .last()
             .map(|bank| bank.transaction_count())
             .unwrap_or(0);
-
-        root_bank.squash();
-        let new_tx_count = root_bank.transaction_count();
-
-        // Generate a snapshot if snapshots are configured and it's been an appropriate number
-        // of banks since the last snapshot
+        // Generate each snapshot at a fixed interval
+        let mut squashed_slot = None;
         if self.snapshot_config.is_some() && snapshot_package_sender.is_some() {
             let config = self.snapshot_config.as_ref().unwrap();
-            info!("setting snapshot root: {}", root);
-            if root - self.last_snapshot_slot >= config.snapshot_interval_slots as Slot {
-                let mut snapshot_time = Measure::start("total-snapshot-ms");
-                let r = self.generate_snapshot(
-                    root,
-                    &root_bank.src.roots(),
-                    snapshot_package_sender.as_ref().unwrap(),
-                );
-                if r.is_err() {
-                    warn!("Error generating snapshot for bank: {}, err: {:?}", root, r);
-                } else {
-                    self.last_snapshot_slot = root;
+            let mut banks = vec![root_bank];
+            let parents = root_bank.parents();
+            banks.extend(parents.iter());
+            for bank in banks.iter() {
+                let root = bank.slot();
+                if bank.block_height() % (config.snapshot_interval_slots as u64) == 0 {
+                    // Generate a snapshot if snapshots are configured and it's been an appropriate number
+                    // of banks since the last snapshot
+                    info!(
+                        "setting snapshot root: {} interval: {}",
+                        root, config.snapshot_interval_slots
+                    );
+                    if root > self.last_snapshot_slot {
+                        bank.squash();
```

Comment:
From https://github.com/solana-labs/solana/pull/8314/files#r385901050

> I'm actually not sure this method of looking at the parents which could be already squashed works, because the old rooted banks could have accounts which are cleaned up and then it would not have the same view into the account set as when it was just recently rooted without cleanup possible.

Is this legimate concern of using parents of squashed bank still a valid one? If so, I think that can apply here as well because now this looping construct is almost identical with the original PR. 

I think this is not valid. In my thought, once `bank` is `squash()`ed, all of their parents are gone and inaccessible, so there should be no chance to loop over the squashed parents in the future call of `set_root()`. Am I missing something?

Reply:
> If so, I think that can apply here as well because now this looping construct is almost identical with the original PR.

Well actually I'm not sure it was a problem in the original code, Since it works slots backwards and only squashes a single slot then it should be safe in only squashing a newer bank than was last squashed. So I agree with you, the concern is not valid.